File: curbcheck/engine/window.py
"""Split a requested parking window into sub-intervals and decide which rules are in force.

A rule's truth value only changes at midnight, at one of its own time
boundaries, or at a calendar boundary (which is a date, hence a midnight). So
we cut [T1, T2] at every distinct boundary the stack under evaluation can
produce and evaluate each piece once. SPEC §9.2 and §9.3.
"""

from __future__ import annotations

from collections.abc import Iterable, Mapping
from dataclasses import dataclass, field
from datetime import date, datetime, timedelta
from itertools import pairwise
from typing import Any

from curbcheck.config import NYC_TZ
from curbcheck.model import Regulation

MINUTES_PER_DAY = 24 * 60


@dataclass(frozen=True)
class Interval:
    """A half-open [start, end) slice of the requested window, never crossing midnight."""

    start: datetime
    end: datetime

    @property
    def weekday(self) -> int:
        """Monday is 0, matching model.Weekday."""
        return self.start.weekday()

    @property
    def date(self) -> date:
        return self.start.date()

    @property
    def minutes(self) -> int:
        """Elapsed real minutes. Subtracting two datetimes in one zone would count wall-clock
        minutes, so the hour that does not happen on the spring-forward Sunday would be billed."""
        return round((self.end.timestamp() - self.start.timestamp()) / 60)

    @property
    def start_minute_of_day(self) -> int:
        return self.start.hour * 60 + self.start.minute


@dataclass(frozen=True)
class CalendarContext:
    """The dates that change what a sign means, as known offline.

    Built from the `asp_suspension` table plus whatever the caller knows about
    school closures. Emergency suspensions are unknowable offline (SPEC §11), so
    an empty context means "nothing special known", never "nothing is happening".
    """

    asp_suspension_dates: frozenset[date] = frozenset()
    major_holiday_dates: frozenset[date] = frozenset()
    meter_suspended_dates: frozenset[date] = frozenset()
    # None means "we have no school calendar"; a school-days rule is then assumed
    # active, because assuming it off would risk a false "legal" (SPEC §11).
    non_school_dates: frozenset[date] | None = None
    snow_emergency: bool = False
    labels: Mapping[date, str] = field(default_factory=dict)

    @classmethod
    def from_asp_rows(
        cls,
        rows: Iterable[Mapping[str, Any]],
        *,
        non_school_dates: Iterable[date] | None = None,
        snow_emergency: bool = False,
    ) -> CalendarContext:
        """Build from `asp_suspension` rows (`date`, `is_major_legal_holiday`, `meters_suspended`, `label`)."""
        suspensions: set[date] = set()
        holidays: set[date] = set()
        meters_off: set[date] = set()
        labels: dict[date, str] = {}
        for row in rows:
            day = date.fromisoformat(str(row["date"]))
            suspensions.add(day)
            if bool(row["is_major_legal_holiday"]):
                holidays.add(day)
            if bool(row["meters_suspended"]):
                meters_off.add(day)
            label = str(row.get("label") or "")
            if label:
                labels[day] = label
        return cls(
            asp_suspension_dates=frozenset(suspensions),
            major_holiday_dates=frozenset(holidays),
            meter_suspended_dates=frozenset(meters_off),
            non_school_dates=None if non_school_dates is None else frozenset(non_school_dates),
            snow_emergency=snow_emergency,
            labels=labels,
        )

    def is_asp_suspended(self, day: date) -> bool:
        return day in self.asp_suspension_dates

    def is_major_holiday(self, day: date) -> bool:
        return day in self.major_holiday_dates

    def is_known_non_school_day(self, day: date) -> bool:
        return self.non_school_dates is not None and day in self.non_school_dates

    def meters_in_effect(self, day: date) -> bool:
        """Meters run every day except Sundays and the six Major Legal Holidays (SPEC §13.2)."""
        if day.weekday() == 6:
            return False
        if day in self.meter_suspended_dates:
            return False
        return not self.is_major_holiday(day)


def expand_window(
    t1: datetime, t2: datetime, regulations: Iterable[Regulation] = ()
) -> list[Interval]:
    """Cut [t1, t2) at midnights and at every rule boundary in `regulations`.

    Both ends must be timezone-aware; parking rules are stated in local time, so
    they are converted to America/New_York first. A window of zero or negative
    length raises.
    """
    start = _as_nyc(t1, "t1")
    end = _as_nyc(t2, "t2")
    if end <= start:
        raise ValueError("the parking window must end after it starts")

    rule_minutes = _rule_boundary_minutes(regulations)
    boundaries = {start, end}
    for day in _dates_covered(start, end):
        for minute in rule_minutes:
            boundaries.add(_at_minute(day, minute))
        boundaries.add(_at_minute(day, MINUTES_PER_DAY))

    ordered = sorted(moment for moment in boundaries if start <= moment <= end)
    return [Interval(left, right) for left, right in pairwise(ordered) if right > left]


def rule_is_active(reg: Regulation, interval: Interval, calendar: CalendarContext) -> bool:
    """Whether the rule is in force for the whole of `interval`.

    Assumes `interval` came from `expand_window` with this rule in the stack, so
    the rule cannot switch on or off part-way through it.
    """
    if interval.weekday not in reg.days:
        return False
    if reg.flags.except_sunday and interval.weekday == 6:
        return False
    if not _in_season(reg, interval.date):
        return False
    if not _in_time_range(reg, interval):
        return False
    return _calendar_allows(reg, interval.date, calendar)


def meter_is_charged(reg: Regulation, interval: Interval, calendar: CalendarContext) -> bool:
    """Whether a metered rule actually costs money in this interval.

    Sundays and Major Legal Holidays are free (SPEC §9.3), except where the sign
    says INCLUDING SUNDAY (§8.4 ex. 10). We read that flag as overriding the
    holiday exemption too: the sign is claiming its own calendar.
    """
    if not reg.metered:
        return False
    if reg.flags.including_sunday:
        return True
    return calendar.meters_in_effect(interval.date)


def _calendar_allows(reg: Regulation, day: date, calendar: CalendarContext) -> bool:
    if reg.flags.street_cleaning and calendar.is_asp_suspended(day):
        return False
    if reg.flags.holiday_exempt and calendar.is_major_holiday(day):
        return False
    if reg.flags.snow_emergency and not calendar.snow_emergency:
        return False
    return not (reg.flags.school_days and calendar.is_known_non_school_day(day))


def _rule_boundary_minutes(regulations: Iterable[Regulation]) -> set[int]:
    minutes = {0}
    for reg in regulations:
        for value in (reg.time_from, reg.time_to):
            if value is not None:
                minutes.add(hhmm_to_minutes(value))
    return minutes


def _dates_covered(start: datetime, end: datetime) -> list[date]:
    days = []
    day = start.date()
    while day <= end.date():
        days.append(day)
        day += timedelta(days=1)
    return days


def _at_minute(day: date, minute_of_day: int) -> datetime:
    """Local wall-clock time on `day`. Minute 1440 is midnight at the start of the next day."""
    midnight = datetime(day.year, day.month, day.day, tzinfo=NYC_TZ)
    return midnight + timedelta(minutes=minute_of_day)


def _as_nyc(moment: datetime, name: str) -> datetime:
    if moment.tzinfo is None or moment.utcoffset() is None:
        raise ValueError(f"{name} must be timezone-aware; parking rules are local time")
    return moment.astimezone(NYC_TZ)


def hhmm_to_minutes(value: str) -> int:
    return int(value[:2]) * 60 + int(value[3:])


def _in_time_range(reg: Regulation, interval: Interval) -> bool:
    if reg.time_from is None or reg.time_to is None:
        return True
    start = hhmm_to_minutes(reg.time_from)
    stop = hhmm_to_minutes(reg.time_to)
    minute = interval.start_minute_of_day
    if stop > start:
        return start <= minute < stop
    # time_to <= time_from wraps past midnight ("8PM-6AM", SPEC §8.4 ex. 15).
    # The equal case is how an all-day range written as 12AM-12AM arrives.
    return minute >= start or minute < stop


def _in_season(reg: Regulation, day: date) -> bool:
    """Seasonal MM-DD bounds are inclusive on both ends and may wrap the year end."""
    if reg.effective_from is None or reg.effective_to is None:
        return True
    today = (day.month, day.day)
    season_start = _mmdd(reg.effective_from)
    season_end = _mmdd(reg.effective_to)
    if season_start <= season_end:
        return season_start <= today <= season_end
    return today >= season_start or today <= season_end


def _mmdd(value: str) -> tuple[int, int]:
    return (int(value[:2]), int(value[3:]))

File: site/static/engine/window.js
/**
 * Split a requested parking window into sub-intervals and decide which rules are in force.
 *
 * A rule's truth value only changes at midnight, at one of its own time
 * boundaries, or at a calendar boundary (which is a date, hence a midnight). So
 * we cut [T1, T2] at every distinct boundary the stack under evaluation can
 * produce and evaluate each piece once. SPEC §9.2 and §9.3.
 *
 * Port of `curbcheck/engine/window.py`; its `CalendarContext` lives in
 * `calendar.js`. Boundaries are New York wall clocks (`../time.js`), because
 * that is what Python compares when two aware datetimes share a tzinfo.
 */

import {
  atMinute,
  compareWall,
  dateKey,
  epochFromWall,
  isWallClock,
  nextDate,
  weekdayOf,
} from "../time.js";

export const MINUTES_PER_DAY = 24 * 60;

/** A half-open [start, end) slice of the requested window, never crossing midnight. */
export class Interval {
  constructor(start, end) {
    this.start = start;
    this.end = end;
    this.startMs = epochFromWall(start);
    this.endMs = epochFromWall(end);
    Object.freeze(this);
  }

  /** Monday is 0, matching model.Weekday. */
  get weekday() {
    return weekdayOf(this.start);
  }

  /** "YYYY-MM-DD", the key `CalendarContext` is looked up by. */
  get date() {
    return dateKey(this.start);
  }

  /**
   * Elapsed real minutes. Subtracting two wall clocks in one zone would count
   * wall-clock minutes, so the hour that does not happen on the spring-forward
   * Sunday would be billed.
   */
  get minutes() {
    return roundHalfEven((this.endMs - this.startMs) / 60_000);
  }

  get startMinuteOfDay() {
    return this.start.hour * 60 + this.start.minute;
  }
}

/**
 * Cut [t1, t2) at midnights and at every rule boundary in `regulations`.
 *
 * Both ends are New York wall clocks; parking rules are stated in local time,
 * and `time.parseRequestTime` has already resolved any offset the caller gave.
 * A window of zero or negative length throws.
 */
export function expandWindow(t1, t2, regulations = []) {
  const start = asNyc(t1, "t1");
  const end = asNyc(t2, "t2");
  if (compareWall(end, start) <= 0) {
    throw new Error("the parking window must end after it starts");
  }

  const ruleMinutes = ruleBoundaryMinutes(regulations);
  // Python collects these in a set of aware datetimes, which deduplicates and
  // orders them by wall clock with `fold` ignored, and keeps the first of any
  // two that compare equal. Insertion order therefore matters: a t1 that
  // arrived as fold=1 has to survive a midnight boundary written as fold=0.
  const boundaries = [start, end];
  for (const day of datesCovered(start, end)) {
    for (const minute of ruleMinutes) {
      boundaries.push(atMinute(day.year, day.month, day.day, minute));
    }
    boundaries.push(atMinute(day.year, day.month, day.day, MINUTES_PER_DAY));
  }

  const inWindow = boundaries.filter(
    (moment) => compareWall(start, moment) <= 0 && compareWall(moment, end) <= 0,
  );
  const ordered = firstOfEachRun(stableSortByWall(inWindow));
  const intervals = [];
  for (let i = 0; i + 1 < ordered.length; i += 1) {
    intervals.push(new Interval(ordered[i], ordered[i + 1]));
  }
  return intervals;
}

/**
 * Whether the rule is in force for the whole of `interval`.
 *
 * Assumes `interval` came from `expandWindow` with this rule in the stack, so
 * the rule cannot switch on or off part-way through it.
 */
export function ruleIsActive(reg, interval, calendar) {
  if (!reg.days.includes(interval.weekday)) {
    return false;
  }
  if (reg.flags.exceptSunday && interval.weekday === 6) {
    return false;
  }
  if (!inSeason(reg, interval.date)) {
    return false;
  }
  if (!inTimeRange(reg, interval)) {
    return false;
  }
  return calendarAllows(reg, interval.date, calendar);
}

/**
 * Whether a metered rule actually costs money in this interval.
 *
 * Sundays and Major Legal Holidays are free (SPEC §9.3), except where the sign
 * says INCLUDING SUNDAY (§8.4 ex. 10). We read that flag as overriding the
 * holiday exemption too: the sign is claiming its own calendar.
 */
export function meterIsCharged(reg, interval, calendar) {
  if (!reg.metered) {
    return false;
  }
  if (reg.flags.includingSunday) {
    return true;
  }
  return calendar.metersInEffect(interval.date);
}

function calendarAllows(reg, day, calendar) {
  if (reg.flags.streetCleaning && calendar.isAspSuspended(day)) {
    return false;
  }
  if (reg.flags.holidayExempt && calendar.isMajorHoliday(day)) {
    return false;
  }
  if (reg.flags.snowEmergency && !calendar.snowEmergency) {
    return false;
  }
  return !(reg.flags.schoolDays && calendar.isKnownNonSchoolDay(day));
}

function ruleBoundaryMinutes(regulations) {
  const minutes = new Set([0]);
  for (const reg of regulations) {
    for (const value of [reg.timeFrom, reg.timeTo]) {
      if (value !== null) {
        minutes.add(hhmmToMinutes(value));
      }
    }
  }
  // Python iterates an unordered set here; every distinct minute-of-day gives a
  // distinct wall clock on a given day, so no two of them can tie and the order
  // never reaches the answer. Sorted anyway so the boundary list is readable.
  return [...minutes].sort((a, b) => a - b);
}

function datesCovered(start, end) {
  const days = [];
  let day = { year: start.year, month: start.month, day: start.day };
  const last = { year: end.year, month: end.month, day: end.day };
  while (compareDate(day, last) <= 0) {
    days.push(day);
    day = nextDate(day);
  }
  return days;
}

function asNyc(moment, name) {
  // `_as_nyc` converts an aware datetime and rejects a naive one. Here every
  // instant has already been resolved to a New York wall clock, so this only
  // rejects the shape `expand_window` would have raised on.
  if (!isWallClock(moment)) {
    throw new Error(`${name} must be timezone-aware; parking rules are local time`);
  }
  return moment;
}

export function hhmmToMinutes(value) {
  return Number(value.slice(0, 2)) * 60 + Number(value.slice(3));
}

function inTimeRange(reg, interval) {
  if (reg.timeFrom === null || reg.timeTo === null) {
    return true;
  }
  const start = hhmmToMinutes(reg.timeFrom);
  const stop = hhmmToMinutes(reg.timeTo);
  const minute = interval.startMinuteOfDay;
  if (stop > start) {
    return start <= minute && minute < stop;
  }
  // time_to <= time_from wraps past midnight ("8PM-6AM", SPEC §8.4 ex. 15).
  // The equal case is how an all-day range written as 12AM-12AM arrives.
  return minute >= start || minute < stop;
}

/** Seasonal MM-DD bounds are inclusive on both ends and may wrap the year end. */
function inSeason(reg, day) {
  if (reg.effectiveFrom === null || reg.effectiveTo === null) {
    return true;
  }
  const today = [Number(day.slice(5, 7)), Number(day.slice(8, 10))];
  const seasonStart = mmdd(reg.effectiveFrom);
  const seasonEnd = mmdd(reg.effectiveTo);
  if (comparePair(seasonStart, seasonEnd) <= 0) {
    return comparePair(seasonStart, today) <= 0 && comparePair(today, seasonEnd) <= 0;
  }
  return comparePair(today, seasonStart) >= 0 || comparePair(today, seasonEnd) <= 0;
}

function mmdd(value) {
  return [Number(value.slice(0, 2)), Number(value.slice(3))];
}

function comparePair([leftMonth, leftDay], [rightMonth, rightDay]) {
  if (leftMonth !== rightMonth) {
    return leftMonth < rightMonth ? -1 : 1;
  }
  if (leftDay !== rightDay) {
    return leftDay < rightDay ? -1 : 1;
  }
  return 0;
}

function compareDate(left, right) {
  if (left.year !== right.year) {
    return left.year < right.year ? -1 : 1;
  }
  return comparePair([left.month, left.day], [right.month, right.day]);
}

function stableSortByWall(moments) {
  // Array.prototype.sort has been stable since ES2019, so equal wall clocks
  // keep the order they were pushed in and the tie goes to the first insert.
  return [...moments].sort(compareWall);
}

function firstOfEachRun(sorted) {
  return sorted.filter(
    (moment, index) => index === 0 || compareWall(sorted[index - 1], moment) < 0,
  );
}

/**
 * Python's `round()` is round-half-to-even, so a window boundary on a half
 * minute would otherwise come out a minute longer here than in the reference.
 */
export function roundHalfEven(value) {
  const floor = Math.floor(value);
  const fraction = value - floor;
  if (fraction > 0.5) {
    return floor + 1;
  }
  if (fraction < 0.5) {
    return floor;
  }
  return floor % 2 === 0 ? floor : floor + 1;
}
